Yoast/duplicate-post
# Tests: make the testsuite compatible with PHPUnit 9 and test against PHP 8

## Context

* Safeguard compatibility with PHP 8

## Summary

This PR can be summarized in the following changelog entry:

* Safeguard compatibility with PHP 8

## Relevant technical choices:

👉 **Pro-tip**: this PR will be easiest to review by going through each commit individually.

### Composer: require Yoast/wp-test-utils

* Adds a dev dependency to the `yoast/wp-test-utils` package, which will also make the PHPUnit Polyfills available.
* As that package already requires and manages the installable versions for BrainMonkey and PHPUnit, remove these as explicit requirements from `require-dev` in favour of letting the WP Test Utils package manage the versions.

This new package adds the following features:
* Polyfills for various PHPUnit cross-version changes.
* Basic test cases for use by the plugins.

Refs:
* https://github.com/Yoast/wp-test-utils
* https://github.com/Yoast/PHPUnit-Polyfills/

### .gitignore: ignore some phpunit related files

* Adds `phpunit.xml` to the `.gitignore` file so overload files from individual developers do not get committed.
* PHPUnit 8 introduces a form of caching to PHPUnit, so adding this cache file to the .gitignore file as well.

### Tests: use the bootstrap file from WP Test Utils

... and remove everything which is already done within that file.

### Tests: switch over to the Yoast\WPTestUtils\BrainMonkey\YoastTestCase

All test classes for the BrainMonkey based unit tests in the Duplicate Post test suite have a Duplicate Post native `TestCase` as a parent class.

This switches the parent of that `TestCase` from the PHPUnit native TestCase to the `Yoast\WPTestUtils\BrainMonkey\YoastTestCase`.

The `YoastTestCase`:
* Inherits PHPUnit cross-version compatibility from the PHPUnit Polyfills TestCase.
* Inherits the BrainMonkey set up and teardown and Mockery expectation counting as assertions from the BrainMonkey TestCase.
* And natively adds a number of stubs for commonly used WP functions.

This switch over includes:
* Removing the addition of function stubs which are already added via the `YoastTestCase` class.
* Removing the `tearDown()` method. This will be inherited from the `YoastTestCase` now.
* Renaming the `setUp()` method to `set_up()` in various test classes for PHPUnit cross-version compatibility.
* Makes the `set_up()` methods `protected` to match the visibility in PHPUnit itself.

There is one "exception", which is the `Post_Duplicator_Test` class.
Various tests in that class set expectations for the WP `sanitize_text_field()` function, which is stubbed in the `YoastTestCase` and as a function can't be both stubbed as well as have expectations set for it, this would lead to failing tests.
So for that particular class, the parent class is switched over to the more generic `Yoast\WPTestUtils\BrainMonkey\TestCase` class, which has the same features as the `YoastTestCase`, with the exception of the additional stubs.

### TestCase: remove the generic stubbing of the translation and escaping functions

Only a limited number of tests - 66 out of 282 - need these stubs, so we may as well only load them if and when needed.

BrainMonkey offers a `stubTranslationFunctions()` function which stubs all nearly translation functions and includes escaping for the "translate and escape" functions.
Similarly, it contains a `stubEscapeFunctions()` function.

The WP Test Utils base class for BrainMonkey offers an alternative implementation for both of these, which doesn't do any escaping, but will always return the original value unchanged.

For each of the 66 tests which actually need either the translation functions or the escaping functions, or both, a call to one or both of the wholesale stubbing functions has been put in place.

As a rule of thumb, these function calls have been added to a test `set_up()` method if more than half the tests in the class need these functions.

Otherwise, the function call(s) have been added to the individual test.

### TestCase: move WP_Role mocking/stubbing to separate function

The mocking of the WP_Role objects and the stubbing of the `get_role()` function is only used by a few tests, so instead of of setting these up for every single test being run, only set these up selectively for those tests which actually use this functionality.

Includes removing the `$roles` property which is unused.

### Tests: remove use of `assertAttributeInstanceOf()` assertions

The `assertAttribute*()` assertion methods were deprecated in PHPUnit 8.x and removed in PHPUnit 9.

The reasoning for that was that `protected` and `private` properties are implementation details and should not be tested directly, but via the methods in the class.

For now, I've elected to maintain the existing test behaviour, while making it work in a PHPUnit cross-version compatible manner.

This is done by:
1. Using a test helper method introduced provided via PHPUnit Polyfills (0.2.0) to get the value of a non-public property.
2. Changing the assertions for the tests from `assertAttribute*()` to `assert*()`

At a later point in time, it should be re-examined whether these implementation details should be tested in this way, if at all.

### Travis: run the tests against PHP 8/nightly

Now the tests are fully compatible with PHPUnit 9, the test suite can be run on PHP 8.

### PHPUnit config: improve code coverage configuration

* Rename the testsuite to use a more descriptive name.
* Improve the code coverage configuration:
    - Add `addUncoveredFilesFromWhitelist` and `processUncoveredFilesFromWhitelist` directives with sensible values.
    - Add the `forceCoversAnnotation` attribute to only record code coverage for the method under test as annotated via a `@covers` annotation.
* Add a separate command in the Composer scripts to run the tests with code coverage.
    Notes:
    - Xdebug with code coverage needs to be enabled for running code coverage.
    - The log preferences will still need to be passed either on the command-line or via a dev-local `phpunit.xml` file for the coverage logging to work.

Based on this configuration, the strict code coverage for this testsuite is currently **59.76%**.

### 🆕 Post_Duplicator_Test: remove expectation for sanitize_text_field() 🆕 

... and extend the normal "base" `TestCase` instead.



## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

* Verify the output of all builds for this PR, making sure that 1) the installation via Composer works without issues, 2) the tests are being run and 3) the tests pass.

To test locally, it will be simplest to:
1. Run `composer update`.
2. Run `composer test` on any PHP version below 8.0 - as this project has no `composer.lock` anymore, probably best to run this initial test run with a low PHP version (5.6/7.0).
    See the tests pass, including a more accurate assertion count (961 compared to 289 previously, due to Mockery expectations now being counted).
    This confirms that the changes to the `TestCase` and the assertion updates to use PHPUnit 9.x syntax works correctly cross-version as on low PHP versions PHPUnit 5.x will be used.
3. Switch to PHP 8.
4. Run `composer update`
5. Run `composer test` to see the tests running and passing on PHP 8 in combination with PHPUnit 9.x.
    This confirms that the tests can now run cross-version on all PHPUnit versions between PHPUnit 5.7 and PHPUnit 9.x and that the tests run and pass on PHP 8.

Note: this doesn't confirm full PHP 8 compatibility yet as the test code coverage as noted above is only ~60%.
